### ISSUE TITLE
Qwen3.8 Quality + Pi: raw CSS enters 8,192-token "!" repetition loop; no guard or tool recovery

### ISSUE THREAD

**Issue by leoli-dev** [youssofal/MTPLX#311]
## Summary

With the MTPLX OpenAI-compatible endpoint and Pi, a normal second tool round on `Youssofal/Qwen3.8-27B-MTPLX-Optimized-Quality` left tool mode, emitted raw CSS into `assistant.content`, then degenerated into a long `!` repetition stream. MTPLX let the stream run to the client-requested 8,192 output-token limit rather than detecting/stopping/recovering it. Pi receives `finish_reason: length` and terminates the agent turn, so the coding task is left incomplete.

This is distinct from #310's tensor-shape crash. It is a reproducible tool-mode / runaway-repetition failure on a server that advertises repetition safeguards.

## Environment

- Hardware: MacBook Pro, Apple M5 Max, 128 GB unified memory
- OS: macOS 26.6.2 (25G83), arm64
- MTPLX app/runtime: 2.9.0 (build 2009001)
- MLX: 0.32.1; mlx-lm: 0.31.3
- Model: `Youssofal/Qwen3.8-27B-MTPLX-Optimized-Quality`
- Served model id: `mtplx-qwen38-27b-optimized-quality`
- Client: Pi 0.84.2, OpenAI-compatible `/v1/chat/completions` provider at `http://127.0.0.1:8001/v1`

Resolved serving configuration:

```text
backend-id qwen3_next
generation-mode mtp
profile turbo
scheduler-mode ar_batch
batching-preset agent
context-window 66560
paged-kv-quantization q8
reasoning-mode off
```

Sampling/client settings at reproduction:

```text
temperature = 1.0
top_p = 0.95
top_k = 20
Pi contextWindow = 66560
Pi maxTokens = 8192
```

## Exact reproduction flow

1. Start MTPLX 2.9.0 with the Quality Qwen3.8 model and its agent/Pi target.
2. Start Pi 0.84.2 with the model above.
3. Ask it to build a normal multi-file HTML/CSS/JS "Arcology 2049" energy-control dashboard: interactive SVG city map, animated UI, controls, chart, modal, etc.
4. Pi makes a valid `write` tool call for `index.html`; MTPLX returns `finish_reason: tool_calls`, and Pi writes the file successfully (`6,206 bytes`).
5. On the immediate next assistant turn, instead of returning a `write` tool call for CSS, the server returns raw CSS text in `assistant.content`.
6. During a CSS declaration it reaches `border:1px solid rgba(74,` and starts emitting `!` repeatedly.
7. The server returns only when the entire 8,192-token output budget is consumed:

```json
{
  "usage": {
    "input": 272,
    "output": 8192,
    "cacheRead": 13754,
    "totalTokens": 22218
  },
  "stopReason": "length",
  "rawStopReason": "length"
}
```

The captured response is 12,067 characters and contains approximately 5,803 exclamation marks. Its tail is entirely repeated `!` characters. Pi then shows:

```text
Response was truncated before completion.
pi-lens deferred format applied to 1 file(s): index.html
```

`pi-lens` is downstream; its formatter was never given a valid CSS file. Pi's behavior after `finish_reason: length` is to stop rather than automatically make an unsafe continuation request, but that does not explain why the endpoint produced thousands of identical tokens.

## Why this looks server-side

- The first round proves the same request/session can produce a valid OpenAI tool call.
- The continuation has a modest effective context (~13,754 cache-read tokens; 22,218 total including the generated failure) under the advertised 66,560-token context window.
- The result is not merely a model preference for CSS prose: it switches from the expected structured tool call to raw content and then emits a pathological single-token loop until max output.
- MTPLX release notes mention safeguards for runaway repetition / whitespace-free repetition, but they did not trigger here for repeated punctuation (`!`).

Temperature 1.0 may make a model-level repetition event more likely, but a local inference/agent server should identify a clearly degenerate stream, cancel it well before max tokens, return an actionable structured error/finish reason, and ideally preserve or recover the tool-call path. Requiring end users to shrink routine coding prompts is not an acceptable fix.

## Expected

1. A standard second agent/tool round should return a valid `write` call for the stylesheet rather than raw partial CSS.
2. Repetition detection should halt a punctuation loop like this quickly, including with MTP verification and `ar_batch` scheduling enabled.
3. The API should return a distinct, actionable finish/error signal for a guarded repetition abort, rather than spending the full `max_tokens` budget and reporting ordinary `length`.
4. Please add a regression test: Qwen3.8 Quality, OpenAI/Pi-style tool transcript, successful HTML `write`, then CSS `write`, with punctuation-repetition detection enabled.
5. Please confirm whether tool serialization/normalization in the OpenAI path can cause the assistant to fall out of tool mode on the continuation.

## Related

- #310: separate Qwen3.8 Pi-tool continuation tensor broadcast crash reported from the same machine/model.
- #282: existing request to make MTPLX tool/system-prompt processing transparent or opt-in.
- #102: previously closed report of an infinite loop with Pi as an MTPLX backend; this reproduction suggests a related class remains in 2.9.0, specifically for a Qwen3.8 MTP tool continuation.

**Comment by youssofal** [comment]
Fixed in the next release: the repetition stop now arms on every request, a client max_tokens is a budget, not a licence to loop, and the same guard now runs per row in batch mode too. Trimmed output is never banked into the session cache and the repetition_stop fields in the response say when it fired. One caveat: tokens already streamed before the guard fires stay on the wire, usage and the non streamed body are trimmed.

Your points about the model leaving tool mode on the continuation are a separate bug and stay open here. Likely mechanism: when the last message is a tool result, the server appends a continuation hint telling the model to answer in plain text, which for "write the stylesheet" reads as an instruction to emit raw CSS. If you can rerun your continuation once as is and once with a trailing user message appended, that would confirm it, the hint only injects when the last message is a tool result.

**Comment by youssofal** [comment]
Fixed in 2.9.2. The literal repetition stop covers batched MTP width 2 and above, which is the lane your CSS case hit.

**Comment by youssofal** [comment]
2.11.3 closes the root cause of the exclamation marks. A non-finite logit used to become token 0, which is `!` in the Qwen vocabulary. The request now ends with `finish_reason: error` and code `non_finite_logits`, the session's cached state is dropped, and the daemon stays up. Release: https://github.com/youssofal/MTPLX/releases/tag/v2.11.3
